processor/metricsgenerationprocessor/utils.go

```diff
@@ -50,36 +51,140 @@ func getMetricValue(metric pmetric.Metric) float64 {
+		toDataPoints = to.Gauge().DataPoints()
+	case pmetric.MetricTypeSum:
+		to.SetEmptySum()
+		metric1DataPoints = metric1.Sum().DataPoints()
+		toDataPoints = to.Sum().DataPoints()
+	default:
+		logger.Debug(fmt.Sprintf("Calculations are only supported on gauge or sum metric types. Given metric '%s' is of type `%s`", metric1.Name(), metricType.String()))
+		return pmetric.NewMetric()
+	}
+
+	// Get metric2 data points
+	var metric2DataPoints pmetric.NumberDataPointSlice
+	switch metricType := metric2.Type(); metricType {
+	case pmetric.MetricTypeGauge:
+		metric2DataPoints = metric2.Gauge().DataPoints()
+	case pmetric.MetricTypeSum:
+		metric2DataPoints = metric2.Sum().DataPoints()
+	default:
+		logger.Debug(fmt.Sprintf("Calculations are only supported on gauge or sum metric types. Given metric '%s' is of type `%s`", metric2.Name(), metricType.String()))
+		return pmetric.NewMetric()
+	}
+
+	for i := 0; i < metric1DataPoints.Len(); i++ {
+		metric1DP := metric1DataPoints.At(i)
+
+		for j := 0; j < metric2DataPoints.Len(); j++ {
+			metric2DP := metric2DataPoints.At(j)
+			if dataPointAttributesMatch(metric1DP, metric2DP) {
+				val, err := calculateValue(dataPointValue(metric1DP), dataPointValue(metric2DP), rule.operation, rule.name)
+
+				if err != nil {
+					logger.Debug(err.Error())
+				} else {
+					newDP := toDataPoints.AppendEmpty()
+					metric1DP.CopyTo(newDP)
+					newDP.SetDoubleValue(val)
+
+					metric2DP.Attributes().Range(func(k string, v pcommon.Value) bool {
+						switch v.Type() {
+						case pcommon.ValueTypeDouble:
+							newDP.Attributes().PutDouble(k, v.Double())
+
+						case pcommon.ValueTypeInt:
+							newDP.Attributes().PutInt(k, v.Int())
+						case pcommon.ValueTypeBool:
+							newDP.Attributes().PutBool(k, v.Bool())
+						default:
+							newDP.Attributes().PutStr(k, v.Str())
+						}
+						// Always return true to ensure iteration over all attributes
+						return true
+					})
 				}
 			}
 		}
 	}
+
+	return to
+}
+
+func dataPointValue(dp pmetric.NumberDataPoint) float64 {
+	switch dp.ValueType() {
+	case pmetric.NumberDataPointValueTypeDouble:
+		return dp.DoubleValue()
+	case pmetric.NumberDataPointValueTypeInt:
+		return float64(dp.IntValue())
+	default:
+		return 0
+	}
+}
+
+func dataPointAttributesMatch(dp1, dp2 pmetric.NumberDataPoint) bool {
+	attributesMatch := true
+	dp1.Attributes().Range(func(key string, dp1Val pcommon.Value) bool {
+		dp1Val.Type()
+		if dp2Val, keyExists := dp2.Attributes().Get(key); keyExists && dp1Val.AsRaw() != dp2Val.AsRaw() {
+			attributesMatch = false
+			return false
+		}
+		return true
+	})
+
+	return attributesMatch
+}
+
+// generateScalarMetrics creates a new metric based on a scalar type rule and adds it to the scope metric.
+// The value for newly calculated metrics is always a floating point number.
+func generateScalarMetrics(rm pmetric.ResourceMetrics, operand2 float64, rule internalRule, logger *zap.Logger) {
```

Comment:
This function is the original `generateMetrics` function, renamed to make the differing behavior obvious.